glean-core/src/metrics/memory_distribution.rs

```diff
@@ -304,4 +305,76 @@ impl MemoryDistributionMetric {
             test_get_num_recorded_errors(glean, self.meta(), error).unwrap_or(0)
         })
     }
+
+    /// Start a new histogram buffer associated with this timing distribution metric.
+    ///
+    /// A histogram buffer accumulates in-memory.
+    /// Data is recorded into the metric on drop.
+    pub fn start_buffer(&self) -> HistogramBuffer<'_> {
+        HistogramBuffer::new(self)
+    }
+
+    fn commit_histogram(&self, histogram: Histogram<Functional>) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| {
+            glean
+                .storage()
+                .record_with(glean, &metric.meta, move |old_value| {
+                    let mut hist = match old_value {
+                        Some(Metric::TimingDistribution(hist)) => hist,
+                        _ => Histogram::functional(LOG_BASE, BUCKETS_PER_MAGNITUDE),
+                    };
+
+                    hist.merge(&histogram);
+                    Metric::TimingDistribution(hist)
+                });
+        });
+    }
+}
+
+/// A histogram buffer associated with a specific instance of a [`MemoryDistributionMetric`].
+///
+/// Accumulation happens in-memory.
+/// Data is merged into the metric on `Drop`.
+#[derive(Debug)]
+pub struct HistogramBuffer<'a> {
+    histogram: Histogram<Functional>,
+    metric: &'a MemoryDistributionMetric,
+}
+
+impl<'a> HistogramBuffer<'a> {
+    /// Create a new histogram buffer referencing the timing distribution it will record into.
+    fn new(metric: &'a MemoryDistributionMetric) -> Self {
+        let histogram = Histogram::functional(LOG_BASE, BUCKETS_PER_MAGNITUDE);
+        Self { histogram, metric }
+    }
+
+    /// Accumulates one sample into the histogram.
+    ///
+    /// The provided sample must be in the "unit" declared by the instance of the metric type
+    /// (e.g. if the instance this method was called on is using [`crate::TimeUnit::Second`], then
+    /// `sample` is assumed to be in seconds).
+    ///
+    /// Accumulation happens in-memory only.
+    pub fn accumulate(&mut self, sample: u64) {
+        self.histogram.accumulate(sample)
+    }
+
+    /// Abandon this histogram buffer and don't commit accumulated data.
+    pub fn abandon(mut self) {
+        // Replace any recordings with an empty histogram.
+        _ = mem::replace(
+            &mut self.histogram,
+            Histogram::functional(LOG_BASE, BUCKETS_PER_MAGNITUDE),
+        );
+    }
+}
+
+impl Drop for HistogramBuffer<'_> {
+    fn drop(&mut self) {
+        // We want to move that value.
+        // A `0/0` histogram doesn't allocate.
+        let buffer = mem::replace(&mut self.histogram, Histogram::functional(0.0, 0.0));
+        self.metric.commit_histogram(buffer);
```

Comment:
We can't `self.metric.commit_histogram(self.histogram);`? We're in `drop`, we can't use any piece of `self` after the method completes.

Reply:
I think I need to do some additional type dance to do it.